### ISSUE TITLE
TestCases Repo Regressions in Travis CI

### ISSUE THREAD

**Issue by economon** [su2code/SU2#459]
All: I have temporarily disabled the regressions for the TestCases repository (the main SU2 repository regressions are still active). The time limit for that repository is only 50 min., as opposed to the 70 min. that we received for the SU2 repo by special request, so it fails essentially every time due to exceeding the limit.

Rather than maintain a separate repository in Travis CI for the TestCases repo, I propose that we implement a way to trigger the main repository regression tests with a change to the TestCases repo. This type of dependent build of downstream repos seems possible by using something like the following:

https://github.com/mernst/plume-lib/blob/master/bin/trigger-travis.sh

What do folks think? Anyone interested in helping with this?

**Comment by economon** [comment]
As a second note: we are aware that even the 70 min. time limit is causing problems for new PRs and other branches that are using Travis CI individually. 

One option is to "parallelize" the builds by decomposing the current regression scripts into more scripts with fewer tests each. However, this only addresses part of the problem, as the compilation step is taking the most time and is currently done within each script. We could consider compiling with lower optimization, etc., to get the cost down, but I am open to ideas on this front too.

**Comment by petebachant** [comment]
How about making the test cases repo a submodule of SU2, such that when the submodule is updated Travis will be triggered?

**Comment by vdweide** [comment]
I don't know all the restrictions of Travis, but wouldn't it be an option to carry the compilation in parallel, i.e. make -j#Procs?

**Comment by economon** [comment]
@petebachant: interesting idea.. do you know if this is possible? I thought that submodules were mostly static links to particular tags of a repo, but this could be a good way.

@vdweide: yes, the machines that Travis CI uses for running open-source projects are fairly small linux boxes with 2-4 cores. Going higher than 'make -j4' won't do much (this is what we do). More ideas along these lines would be great though.

By the way, you can see the details of how we download the code, packages, and compile/install in the .travis.yml file here:

https://github.com/su2code/SU2/blob/develop/.travis.yml

**Comment by petebachant** [comment]
I think it is possible, but there may need to be some renaming to deal with the fact that there's already a `TestCases` directory. One benefit of using a submodule is that an exact commit of the test cases repo can be tied to each SU2 commit. Also, I think Travis by default recurses submodules when cloning.

**Comment by economon** [comment]
The short term fix for reducing the Travis build time for the regressions is working well, but we need a longer term solution. A couple more ideas for discussion...

1. See if we can take advantage of the caching capability in Travis to save build time: https://docs.travis-ci.com/user/caching/#How-does-the-caching-work. Anyone familiar with this already?

2. More work, but probably the best long term solution: divide some files, especially the large header files, and be better about our #include statements. Sometime in the future, we could even pursue custom executable builds for a certain solver (flow, structure, etc.) that only require subsets of files.

**Comment by talbring** [comment]
The caching feature actually sounds very useful. I'll have a look into it.
